src/socket_can_receiver_node.cpp

```diff
@@ -103,22 +101,25 @@ void SocketCanReceiverNode::receive()
   CanId receive_id{};
   can_msgs::msg::Frame frame_msg(rosidl_runtime_cpp::MessageInitialization::ZERO);
   frame_msg.header.frame_id = "can";
-  try {
-    receive_id = receiver_->receive(frame_msg.data.data(), interval_ns_);
-  } catch (const std::exception & ex) {
-    RCLCPP_WARN_THROTTLE(
-      this->get_logger(), *this->get_clock(), 1000,
-      "Error receiving CAN message: %s - %s",
-      interface_.c_str(), ex.what());
-    return;
+
+  while (rclcpp::ok()) {
+    try {
+      receive_id = receiver_->receive(frame_msg.data.data(), interval_ns_);
+    } catch (const std::exception & ex) {
+      RCLCPP_WARN_THROTTLE(
+        this->get_logger(), *this->get_clock(), 1000,
+        "Error receiving CAN message: %s - %s",
+        interface_.c_str(), ex.what());
+      continue;
+    }
+    frame_msg.header.stamp = this->now();
+    frame_msg.id = receive_id.identifier();
+    frame_msg.is_rtr = (receive_id.frame_type() == FrameType::REMOTE);
+    frame_msg.is_extended = receive_id.is_extended();
+    frame_msg.is_error = (receive_id.frame_type() == FrameType::ERROR);
+    frame_msg.dlc = receive_id.length();
+    frames_pub_->publish(std::move(frame_msg));
```

Comment:
This must have at least *some* sleep time - even if it is 500 ns. If there is no sleep time, this thread will take 100% utilization of one CPU core. I am fine running it as a separate thread but it must not run an infinite loop with no sleep.

Additionally (which I forgot), this should only attempt to receive messages if `this->get_current_state() == State::ACTIVE` since the thread is created in the `on_configure` callback but we only want to actually receive messages when in the `ACTIVE` state. The publisher will only publish while in the `ACTIVE` state but it is still better to only attempt to read messages when in `ACTIVE` as well.

Reply:
@JWhitleyWork I've tested the CPU usage, but it was very low and not using 100% of 1 core.

![image](https://user-images.githubusercontent.com/31987104/112247541-d10a5e00-8c97-11eb-9ffd-0755cd57736b.png)

https://user-images.githubusercontent.com/31987104/112247573-df587a00-8c97-11eb-9908-50e0c2da5395.mp4

I guess `select` here affects?
https://github.com/autowarefoundation/ros2_socketcan/blob/468719fc73d0d8bd8387bd2e61a3ecabac55c2bb/src/socket_can_receiver.cpp#L53

I'm okay to simply add `sleep`, but I think it's not good to add it if it's not necessary. What do you think?

Reply:
Thanks for looking into this! I'm OK leaving it without a sleep but the check for the `ACTIVE` status should still be added.

Reply:
Yes, I'm working on it right now!

Reply:
@JWhitleyWork Oh, I found when I deactivate the receiver, there are some problems. I'll fix these.

1. Before checking ACTIVE

It spams to the console.

```sh
[socket_can_receiver_node_exe-1] [WARN 1616557238.722636473] [LifecyclePublisher]: Trying to publish message on the topic '/from_can_bus', but the publisher is not activated
[socket_can_receiver_node_exe-1] [WARN 1616557238.734286816] [LifecyclePublisher]: Trying to publish message on the topic '/from_can_bus', but the publisher is not activated
[socket_can_receiver_node_exe-1] [WARN 1616557238.734306586] [LifecyclePublisher]: Trying to publish message on the topic '/from_can_bus', but the publisher is not activated
[socket_can_receiver_node_exe-1] [WARN 1616557238.734312886] [LifecyclePublisher]: Trying to publish message on the topic '/from_can_bus', but the publisher is not activated
```

2. After checking ACTIVE

No spamming, but CPU usage is 100% because `select` isn't called anymore.

Reply:
@JWhitleyWork Fixed! Could you check again?

https://user-images.githubusercontent.com/31987104/112251811-40d01700-8c9f-11eb-924c-7c8e3a286f1a.mp4

I concern about a bit high CPU usage when deactivated. (activated: 2%, deactivated: 4%)